File: tests/fixtures/core-identity.ts
import type {CoreEntity, CoreEntityKind, CoreIdentityGraph, CoreRelation, CoreRef} from "../../packages/contracts/src/spatial/core";

export const ref = (id:string,namespace="physical"):CoreRef => ({namespace,id});
export function entity(id:string,kind:CoreEntityKind="building",namespace="physical"):CoreEntity {
  return {ref:ref(id,namespace),revision:1,kind,label:`Object ${id}`,identifiers:[],memberships:[],lifecycle:{state:"active"}};
}
export function relation(id:string,kind:CoreRelation["kind"],from:CoreRef,to:CoreRef):CoreRelation {
  return {id,revision:1,kind,from,to,note:"Explicit authored fixture relationship"};
}
export function identityFixture():CoreIdentityGraph {
  const b1={...entity("B1"),label:"Same label",identifiers:[{scheme:"source_property_id",issuer:"Fixture issuer",value:"0000123",status:"reported" as const,historical:false}],memberships:[{collection:ref("west","area"),role:"authoring" as const},{collection:ref("east","area"),role:"authoring" as const}]};
  const b2={...entity("B2"),label:"Same label",identifiers:b1.identifiers};
  return {entities:[b1,b2,entity("L1","level"),entity("L2","level"),entity("U1","space"),entity("stairs","space"),entity("P1","parcel"),entity("B1","building","registry"),entity("pipe","utility"),entity("road","road")],relations:[
    relation("level-1","part_of",ref("L1"),ref("B1")),relation("level-2","part_of",ref("L2"),ref("B1")),
    relation("duplex-parent","part_of",ref("U1"),ref("B1")),relation("duplex-lower","occupies_level",ref("U1"),ref("L1")),relation("duplex-upper","occupies_level",ref("U1"),ref("L2")),
    relation("stairs-west","part_of",ref("stairs"),ref("B1")),relation("stairs-east","part_of",ref("stairs"),ref("B2")),
    relation("parcel-link","associated_parcel",ref("B1"),ref("P1")),relation("physical-record","recorded_by",ref("B1"),ref("B1","registry")),
    relation("pipe-service","serves",ref("pipe"),ref("B1")),relation("pipe-crossing","crosses",ref("pipe"),ref("road")),
  ]};
}
export function splitFixture():CoreIdentityGraph {
  return {entities:[{...entity("P0","parcel"),revision:2,lifecycle:{state:"retired",mode:"split",changeId:"split-1",replacedBy:[ref("P1"),ref("P2")],reason:"Explicit identity split, no geometry inferred"}},entity("P1","parcel"),entity("P2","parcel")],relations:[
    {...relation("split-1:a","split_from",ref("P1"),ref("P0")),changeId:"split-1"},
    {...relation("split-1:b","split_from",ref("P2"),ref("P0")),changeId:"split-1"},
  ]};
}
export function mergeFixture():CoreIdentityGraph {
  return {entities:[...['P1','P2'].map(id=>({...entity(id,"parcel"),revision:2,lifecycle:{state:"retired" as const,mode:"merge" as const,changeId:"merge-1",replacedBy:[ref("P3")],reason:"Explicit identity merge"}})),entity("P3","parcel")],relations:[
    {...relation("merge-1:a","merged_from",ref("P3"),ref("P1")),changeId:"merge-1"},
    {...relation("merge-1:b","merged_from",ref("P3"),ref("P2")),changeId:"merge-1"},
  ]};
}
export interface IdentityConformanceCase {id:string;value:unknown;valid:boolean;code?:string}
export function identityCases():IdentityConformanceCase[] {
  const cases:IdentityConformanceCase[]=[{id:"empty-catalog",value:{entities:[],relations:[]},valid:true},{id:"duplex-shared-stairs-and-namespaces",value:identityFixture(),valid:true},{id:"explicit-split-history",value:splitFixture(),valid:true},{id:"explicit-merge-history",value:mergeFixture(),valid:true}];
  const change=(id:string,code:string,mutate:(value:any)=>void,base:()=>unknown=identityFixture)=>{const value=JSON.parse(JSON.stringify(base()));mutate(value);cases.push({id,value,valid:false,code});};
  change("duplicate-reference","DUPLICATE_ENTITY",g=>g.entities.push(g.entities[0]));
  change("duplicate-membership","DUPLICATE_MEMBERSHIP",g=>g.entities[0].memberships.push(g.entities[0].memberships[0]));
  change("duplicate-identifier-assertion","DUPLICATE_IDENTIFIER",g=>g.entities[0].identifiers.push({...g.entities[0].identifiers[0],historical:true}));
  change("parcel-identifier-on-building","IDENTIFIER_SCOPE",g=>g.entities[0].identifiers[0].scheme="official_ulpin");
  change("numeric-external-identifier","INVALID_CONTRACT",g=>g.entities[0].identifiers[0].value=123);
  change("fractional-revision","INVALID_CONTRACT",g=>g.entities[0].revision=1.5);
  change("boolean-revision","INVALID_CONTRACT",g=>g.entities[0].revision=true);
  change("unsafe-integer-revision","INVALID_CONTRACT",g=>g.entities[0].revision=9007199254740992);
  change("unrecognized-kind","INVALID_CONTRACT",g=>g.entities[0].kind="apartment-guessed");
  change("unknown-core-field","INVALID_CONTRACT",g=>g.entities[0].residents=['private']);
  change("control-in-key","INVALID_CONTRACT",g=>g.entities[0].ref.id="B1\n");
  change("missing-relation-endpoint","MISSING_ENDPOINT",g=>g.relations[0].to=ref("absent"));
  change("wrong-relation-kinds","RELATION_KIND",g=>g.relations[0].kind="occupies_level");
  change("same-reference-relation","SELF_RELATION",g=>g.relations[0].to=g.relations[0].from);
  change("duplicate-edge-id","DUPLICATE_RELATION_ID",g=>g.relations[1].id=g.relations[0].id);
  change("duplicate-directed-edge","DUPLICATE_RELATION",g=>g.relations.push({...g.relations[0],id:"duplicate-edge"}));
  change("reverse-crossing-not-new-edge","DUPLICATE_RELATION",g=>g.relations.push(relation("duplicate-crossing","crosses",ref("road"),ref("pipe"))));
  change("recording-in-physical-namespace","RECORD_NAMESPACE",g=>g.relations.find((r:any)=>r.kind==="recorded_by").to=ref("B2"));
  change("non-lineage-change-id","LINEAGE_CHANGE",g=>g.relations[0].changeId="unexpected");
  change("containment-cycle","RELATION_CYCLE",g=>{g.entities=[entity("a","building_part"),entity("b","building_part")];g.relations=[relation("a-b","part_of",ref("a"),ref("b")),relation("b-a","part_of",ref("b"),ref("a"))];});
  change("missing-successor","MISSING_SUCCESSOR",g=>g.entities[0].lifecycle.replacedBy[0]=ref("missing"),splitFixture);
  change("duplicate-successor","DUPLICATE_SUCCESSOR",g=>g.entities[0].lifecycle.replacedBy[1]=ref("P1"),splitFixture);
  change("single-child-split","SUCCESSOR_COUNT",g=>g.entities[0].lifecycle.replacedBy.pop(),splitFixture);
  change("successor-kind-change","INVALID_SUCCESSOR",g=>g.entities[1].kind="building",splitFixture);
  change("lineage-change-mismatch","LINEAGE_CHANGE",g=>g.relations[0].changeId="wrong",splitFixture);
  change("one-source-merge","RETIREMENT_GROUP",g=>{g.entities.splice(1,1);g.relations.splice(1,1);},mergeFixture);
  change("multiple-target-merge","SUCCESSOR_COUNT",g=>g.entities[0].lifecycle.replacedBy.push(ref("P2")),mergeFixture);
  change("reserved-prototype-key","UNSAFE_KEY",g=>Object.defineProperty(g,"__proto__",{value:{polluted:true},enumerable:true}));
  const reordered=identityFixture();cases.push({id:"source-order-independent",value:{entities:[...reordered.entities].reverse(),relations:[...reordered.relations].reverse()},valid:true});
  return cases;
}

File: services/geo/geo/core_identity.py
"""Identity metadata semantics shared with TypeScript via generated policy/corpus."""
from __future__ import annotations
from .core_contract import bundled_schema, fail, parse_core, ref_key


def unique(values, code):
    if len(set(values)) != len(values):
        fail(code, "Duplicate references/assertions are not permitted")


def acyclic(edges, group):
    outgoing, indegree = {}, {}
    for source, target in edges:
        outgoing.setdefault(source, []).append(target)
        indegree.setdefault(source, 0)
        indegree[target] = indegree.get(target, 0) + 1
    ready = [key for key, count in indegree.items() if count == 0]
    cursor = 0
    while cursor < len(ready):
        for target in outgoing.get(ready[cursor], []):
            indegree[target] -= 1
            if indegree[target] == 0:
                ready.append(target)
        cursor += 1
    if cursor != len(indegree):
        fail("RELATION_CYCLE", "Cycle in " + group + " relationships")


def validate_core_identity_graph(value):
    graph = parse_core("identity-graph", value)
    policy = bundled_schema("identity-graph")[0]["x-ulpin-relation-policy"]
    entities = {}
    for entity in graph["entities"]:
        key = ref_key(entity["ref"])
        if key in entities:
            fail("DUPLICATE_ENTITY", "A canonical reference appears more than once")
        entities[key] = entity
        unique([(ref_key(m["collection"]), m["role"]) for m in entity["memberships"]], "DUPLICATE_MEMBERSHIP")
        unique([(i["scheme"], i["issuer"], i["value"]) for i in entity["identifiers"]], "DUPLICATE_IDENTIFIER")
        if any(i["scheme"] == "official_ulpin" for i in entity["identifiers"]) and entity["kind"] != "parcel":
            fail("IDENTIFIER_SCOPE", "A parcel identifier cannot become a building identifier")
    replacement_edges, changes, allocation_changes = [], {}, {}
    for entity in graph["entities"]:
        lifecycle = entity["lifecycle"]
        if lifecycle["state"] != "retired":
            continue
        source = ref_key(entity["ref"])
        targets = [ref_key(r) for r in lifecycle["replacedBy"]]
        unique(targets, "DUPLICATE_SUCCESSOR")
        if (lifecycle["mode"] == "split" and len(targets) < 2) or (lifecycle["mode"] == "merge" and len(targets) != 1):
            fail("SUCCESSOR_COUNT", "Retirement successors must match the operation")
        for target_key in targets:
            target = entities.get(target_key)
            if target is None:
                fail("MISSING_SUCCESSOR", "A retired identity's successor is missing")
            if target_key == source or target["kind"] != entity["kind"] or target["ref"]["namespace"] != entity["ref"]["namespace"]:
                fail("INVALID_SUCCESSOR", "Identity successor kind or namespace differs")
            if target_key in allocation_changes and allocation_changes[target_key] != lifecycle["changeId"]:
                fail("SUCCESSOR_ALLOCATION", "A successor cannot be allocated by unrelated changes")
            allocation_changes[target_key] = lifecycle["changeId"]
            replacement_edges.append((source, target_key))
        changes.setdefault(lifecycle["changeId"], []).append(entity)
    acyclic(replacement_edges, "replacement")
    for group in changes.values():
        first = group[0]["lifecycle"]
        if any(e["lifecycle"]["mode"] != first["mode"] for e in group):
            fail("RETIREMENT_GROUP", "One change cannot mix split and merge")
        if first["mode"] == "split" and len(group) != 1:
            fail("RETIREMENT_GROUP", "A split identifies one original")
        if first["mode"] == "merge" and (len(group) < 2 or any(ref_key(e["lifecycle"]["replacedBy"][0]) != ref_key(first["replacedBy"][0]) for e in group)):
            fail("RETIREMENT_GROUP", "A merge identifies at least two originals and one successor")
    unique([r["id"] for r in graph["relations"]], "DUPLICATE_RELATION_ID")
    keys, groups, lineage_keys = [], {}, set()
    for relation in graph["relations"]:
        source_key, target_key = ref_key(relation["from"]), ref_key(relation["to"])
        source, target = entities.get(source_key), entities.get(target_key)
        if source is None or target is None:
            fail("MISSING_ENDPOINT", "A relationship endpoint is missing")
        if source_key == target_key:
            fail("SELF_RELATION", "A relationship cannot target itself")
        rule = policy[relation["kind"]]
        if [source["kind"], target["kind"]] not in rule["pairs"]:
            fail("RELATION_KIND", "Relationship endpoint kinds are incompatible")
        if relation["kind"] == "recorded_by" and (target["ref"]["namespace"] != "registry" or source["ref"]["namespace"] == "registry"):
            fail("RECORD_NAMESPACE", "Recording must explicitly target a distinct registry identity")
        if relation["kind"] in ("split_from", "merged_from"):
            if source["ref"]["namespace"] != target["ref"]["namespace"]:
                fail("LINEAGE_NAMESPACE", "Lineage cannot change namespace")
            lifecycle = target["lifecycle"]
            mode = "split" if relation["kind"] == "split_from" else "merge"
            if lifecycle["state"] != "retired" or lifecycle["changeId"] != relation.get("changeId") or lifecycle["mode"] != mode or source_key not in [ref_key(r) for r in lifecycle["replacedBy"]]:
                fail("LINEAGE_CHANGE", "Lineage must agree with the recorded retirement")
            lineage_keys.add((relation["kind"], source_key, target_key, relation["changeId"]))
        elif "changeId" in relation:
            fail("LINEAGE_CHANGE", "Only lineage can carry an identity change ID")
        pair = sorted((source_key, target_key)) if rule["symmetric"] else (source_key, target_key)
        keys.append((relation["kind"], *pair))
        if rule["cycleGroup"]:
            groups.setdefault(rule["cycleGroup"], []).append((source_key, target_key))
    unique(keys, "DUPLICATE_RELATION")
    for group, edges in groups.items():
        acyclic(edges, group)
    for entity in graph["entities"]:
        lifecycle = entity["lifecycle"]
        if lifecycle["state"] == "retired":
            kind = "split_from" if lifecycle["mode"] == "split" else "merged_from"
            for successor in lifecycle["replacedBy"]:
                if (kind, ref_key(successor), ref_key(entity["ref"]), lifecycle["changeId"]) not in lineage_keys:
                    fail("LINEAGE_MISSING", "Every successor requires explicit lineage")
    return graph
